File: shared/event_bus.py
# shared/event_bus.py

from collections import defaultdict
from uuid import uuid4

class EventBus:
    def __init__(self):
        self.subscribers = {}

    def subscribe(self, topic, callback):
        subscription_id = f"sub_{len(self.subscribers)+1}"
        self.subscribers[subscription_id] = {'topic': topic, 'callback': callback}
        return subscription_id

    def publish(self, topic, payload):
        success = False
        for sub in self.subscribers.values():
            if sub['topic'] == topic:
                sub['callback'](payload)
                success = True
        return success

    def unsubscribe(self, subscription_id):
        return self.subscribers.pop(subscription_id, None) is not None

File: shared/event_bus.js
// Usage: import EventBus from 'shared/event_bus.js';
// EventBus.subscribe('topic', callback);   // returns subscriptionId
// EventBus.publish('topic', payload);      // returns true if successful
class EventBus {
  constructor() {
    this.subscribers = {};
  }

  subscribe(topic, callback) {
    const subscriptionId = `sub_${Object.keys(this.subscribers).length + 1}`;
    this.subscribers[subscription_id] = { topic, callback };
    return subscription_id;
  }

  publish(topic, payload) {
    let success = false;
    Object.values(this.subscribers).forEach(sub => {
      if (sub.topic === topic) {
        sub.callback(payload);
        success = true;
      }
    });
    return success;
  }

  unsubscribe(subscriptionId) {
    if (this.subscribers[subscriptionId]) {
      delete this.subscribers[subscriptionId];
      return true;
    }
    return false;
  }
}

export default new EventBus();
